dynamic chunk sizing for v4 raw forward index

itschrispeck opened the pull request:
**Background**
- V4 format was introduced to better handle variable length data chunk size by reducing the potential for large allocations https://github.com/apache/pinot/pull/7661 
- V3 format allocated direct memory based on `numDocPerChunk * lengthOfLongestEntry`, which was very efficient for near-constant length/short data. 
- For example, in each format a null column’s chunk size would be:
  - V3: 1000 docs * 4 bytes (‘null’) = `4KB`
  - V4: `1MB` hardcoded target

**Problem**
- Making V4 default (https://github.com/apache/pinot/pull/11120) will result in a large direct memory increase for the values we typically see. 
- For the static 1000 docs/chunk used in V2/3 (assuming `deriveNumDocsPerChunk` is not set) the breakeven point assumes the `lengthOfLongestEntry` of a column in a segment is ~1KB

We have seen this behavior first hand after making V4 default internally. We have many columns for which we do not know if they will contain variable length data or ‘short data’, and it's desirable to handle both cases with a single format.

**Change**
This PR introduces dynamic chunk sizing for V4 format. Target chunk size is calculated based on the heuristic:
```
max(min(maxLength * targetDocsPerChunk, targetMaxChunkSize), TARGET_MIN_CHUNK_SIZE)
```
where new configs are introduced:
```
"forward": {
  "targetMaxChunkSize": "1M",
  "targetDocsPerChunk": 1000
}
```
and `TARGET_MIN_CHUNK_SIZE = 4K`

In testing I’ve found doing this results in reduced direct memory spikes, especially against wide tables/high QPS. The below graph shows the improvement in direct memory spikes for a env with majority of tables using 3-7 day TTL and adhoc QPS. Some spikes are still present as not all segments with the old static chunk size have been expired (some 30 day TTL tables exist).

![image](https://github.com/apache/pinot/assets/27231838/b414a7a3-240a-4d76-8a01-987e552b1b1a)

I think dynamic chunk sizing should be the default implementation for V4 and have not put this behind a config. It bridges the gap between the variable length data behavior of V4 with the 'short data' behavior of V2/V3. 

There are no backward compatibility concerns with this PR. 

tags: `performance`

itschrispeck commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/SingleValueVarByteRawIndexCreator.java:
```diff
@@ -38,6 +38,7 @@
  */
 public class SingleValueVarByteRawIndexCreator implements ForwardIndexCreator {
   private static final int DEFAULT_NUM_DOCS_PER_CHUNK = 1000;
+  private static final int TARGET_MIN_CHUNK_SIZE = 4 * 1024;
```
This lower bound is debatable. 4KB is what we tested with, but since it's uncompressed target size it could be below disk read ahead value for many systems

Jackie-Jiang commented:
I don't think we need a lower bound for the chunk size. We can probably simply do `min(maxLength * DEFAULT_NUM_DOCS_PER_CHUNK, TARGET_MAX_CHUNK_SIZE)`.
I feel it can also be useful to allow user to specify the chunk size

itschrispeck commented:
> I don't think we need a lower bound for the chunk size. We can probably simply do `min(maxLength * DEFAULT_NUM_DOCS_PER_CHUNK, TARGET_MAX_CHUNK_SIZE)`.

I had the same thought, but we ran into two issues that blocked segment build: maxLength can be 0, and int overflow for large maxLength. Setting a minimum size seemed like a good way to catch both cases. 

> I feel it can also be useful to allow user to specify the chunk size

Makes a lot of sense. I added a config `targetMaxChunkSize` which sets the upper bound. The chunk size can still be dynamically reduced if `maxLength` is small, since I couldn't think of a strong case for a user increasing chunk size when values are always short. I will document the behavior in the docs. Reducing the max chunk size is very useful for both avoiding on the fly allocations/huge chunks w/ V4, and in reducing direct buffer usage.

This config can also apply to V2/V3 format with `deriveNumDocsPerChunk`.

Jackie-Jiang commented:
`DEFAULT_NUM_DOCS_PER_CHUNK` is the thing we want to avoid in V4, so using it to calculate the target chunk size seems weird to me. How about we add 2 configs here:
- `maxTargetChunkSize`: upper bound of the target chunk size
- `targetDocsPerChunk`: reduce the target chunk size when max length is small. We can make it 1000 by default to have the desired behavior. I can imagine people want to disable this and always go with the `maxTargetChunkSize` for scan intensive case to reduce decompression

richardstartin commented:
Nice improvement!

itschrispeck commented:
> How about we add 2 configs here

Done. Updated the description with the new configs. Lmk if there's any concern with naming, I changed it slightly for consistency: 
```
"forward": {
  "targetMaxChunkSize": "1M",
  "targetDocsPerChunk": 1000
}
```


Posting the below just for reference: 
>I can imagine people want to disable this and always go with the maxTargetChunkSize for scan intensive case to reduce decompression 

I understand why this should be the case, but could not reproduce it in our prod env. I ended up being able to reproduce it in a microbench, and it turned out the scans I initially tested were against very low cardinality data. For higher cardinality data the microbench showed up to 40% faster performance w/ 1M vs 4K chunk size.

Jackie-Jiang commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/ForwardIndexConfig.java:
```diff
@@ -31,30 +31,46 @@
+    }
+    _targetMaxChunkSize =
+        targetMaxChunkSize == null ? DataSizeUtils.fromBytes(DEFAULT_TARGET_MAX_CHUNK_SIZE) : targetMaxChunkSize;
+    _targetDocsPerChunk =
```
Suggest allowing negative value for this config to turn it off and only honor the max chunk size.

Jackie-Jiang reviewed (approved):
LGTM. Can you help also update the pinot documentation about these 2 configs? Some examples will help user understand how to use them.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/SingleValueVarByteRawIndexCreator.java:
```diff
@@ -70,23 +69,32 @@ public SingleValueVarByteRawIndexCreator(File baseIndexDir, ChunkCompressionType
+
+    // For columns with very small max value, target chunk size should also be capped to reduce memory during read
+    int dynamicTargetChunkSize =
+        Math.max(Math.min(maxLength * targetDocsPerChunk, targetMaxChunkSizeBytes), TARGET_MIN_CHUNK_SIZE);
```
I guess this will overflow when `targetDocsPerChunk` is `Integer.MAX_VALUE`. Let's add a test to ensure overflow is properly handled

itschrispeck replied:
thanks for the catch!

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/ForwardIndexUtils.java:
```diff
@@ -0,0 +1,42 @@
+   */
+  public static int getDynamicTargetChunkSize(int maxLength, int targetDocsPerChunk, int targetMaxChunkSizeBytes) {
+    if (targetDocsPerChunk < 0 || (long) maxLength * targetDocsPerChunk > Integer.MAX_VALUE) {
+      return targetMaxChunkSizeBytes;
```
Maybe also put a lower bound to this?

Jackie-Jiang commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/ForwardIndexConfig.java:
```diff
@@ -31,30 +31,45 @@
+          "targetMaxChunkSize should only be used when deriveNumDocsPerChunk is true or rawIndexWriterVersion is 4");
+    }
+    _targetMaxChunkSize =
+        targetMaxChunkSize == null ? DataSizeUtils.fromBytes(DEFAULT_TARGET_MAX_CHUNK_SIZE) : targetMaxChunkSize;
```
(minor) Make a constant for `DataSizeUtils.fromBytes(DEFAULT_TARGET_MAX_CHUNK_SIZE)` (maybe having both `DEFAULT_TARGET_MAX_CHUNK_SIZE` and `DEFAULT_TARGET_MAX_CHUNK_SIZE_BYTES`)

Jackie-Jiang commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/ForwardIndexConfig.java:
```diff
@@ -31,30 +31,45 @@
   private final CompressionCodec _compressionCodec;
   private final boolean _deriveNumDocsPerChunk;
   private final int _rawIndexWriterVersion;
+  private final String _targetMaxChunkSize;
```
Consider parsing `_targetMaxChunkSizeBytes` upfront to avoid illegal size

klsince commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/fwd/SingleValueVarByteRawIndexCreator.java:
```diff
@@ -70,23 +68,32 @@ public SingleValueVarByteRawIndexCreator(File baseIndexDir, ChunkCompressionType
+
+    // For columns with very small max value, target chunk size should also be capped to reduce memory during read
+    int dynamicTargetChunkSize =
+        ForwardIndexUtils.getDynamicTargetChunkSize(maxLength, targetDocsPerChunk, targetMaxChunkSizeBytes);
```
should this method take numDocsPerChunk instead of targetDocsPerChunk here?

Jackie-Jiang replied:
I think it is correct. If not configured, `targetDocsPerChunk` should be 1000 by default.
Made a small cleanup PR #13093 to clarify the logic a little bit
